src/awkward/_v2/operations/structure/ak_flatten.py

```diff
@@ -8,163 +8,153 @@
+                  [],
+                  [
+                   [7.7],
+                   [8.8, 9.9]
+                  ]])
+
+    At `axis=1`, the outer lists (length 4, length 0, length 2) become a single
+    list (of length 6).
+
+        >>> print(ak.flatten(array, axis=1))
+        [[1.1, 2.2, 3.3], [], [4.4, 5.5], [6.6], [7.7], [8.8, 9.9]]
+
+    At `axis=2`, the inner lists (lengths 3, 0, 2, 1, 1, and 2) become three
+    lists (of lengths 6, 0, and 3).
+
+        >>> print(ak.flatten(array, axis=2))
+        [[1.1, 2.2, 3.3, 4.4, 5.5, 6.6], [], [7.7, 8.8, 9.9]]
+
+    There's also an option to completely flatten the array with `axis=None`.
+    This is useful for passing the data to a function that doesn't care about
+    nested structure, such as a plotting routine.
+
+        >>> print(ak.flatten(array, axis=None))
+        [1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]
+
+    Missing values are eliminated by flattening: there is no distinction
+    between an empty list and a value of None at the level of flattening.
+
+        >>> array = ak.Array([[1.1, 2.2, 3.3], None, [4.4], [], [5.5]])
+        >>> ak.flatten(array, axis=1)
+        <Array [1.1, 2.2, 3.3, 4.4, 5.5] type='5 * float64'>
+
+    As a consequence, flattening at `axis=0` does only one thing: it removes
+    None values from the top level.
+
+        >>> ak.flatten(array, axis=0)
+        <Array [[1.1, 2.2, 3.3], [4.4], [], [5.5]] type='4 * var * float64'>
+
+    As a technical detail, the flattening operation can be trivial in a common
+    case, #ak.layout.ListOffsetArray in which the first `offset` is `0`.
+    In that case, the flattened data is simply the array node's `content`.
+
+        >>> array.layout
+        <ListOffsetArray64>
+            <offsets><Index64 i="[0 4 4 6]" offset="0" length="4"/></offsets>
+            <content><ListOffsetArray64>
+                <offsets><Index64 i="[0 3 3 5 6 7 9]" offset="0" length="7"/></offsets>
+                <content>
+                    <NumpyArray format="d" shape="9" data="1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9"/>
+                </content>
+            </ListOffsetArray64></content>
+        </ListOffsetArray64>
+        >>> np.asarray(array.layout.content.content)
+        array([1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9])
+
+    However, it is important to keep in mind that this is a special case:
+    #ak.flatten and `content` are not interchangeable!
+    """
+    layout = ak._v2.operations.convert.to_layout(
+        array, allow_record=False, allow_other=False
+    )
+    nplike = ak.nplike.of(layout)
+
+    if axis is None:
+        out = (layout.completely_flatten(),)
+
+        assert isinstance(out, tuple) and all(isinstance(x, np.ndarray) for x in out)
+
+        if any(isinstance(x, nplike.ma.MaskedArray) for x in out):
+            out = ak._v2.contents.NumpyArray(nplike.ma.concatenate(out))
+        else:
+            out = ak._v2.contents.NumpyArray(nplike.concatenate(out))
```

Comment:
v1 made this same mistake:

https://github.com/scikit-hep/awkward-1.0/blob/4071495c93fe50443827bbc286ace45ae5fa7ba0/src/awkward/operations/structure.py#L1903

a NumPy MaskedArray can't be ingested directly into a NumpyArray; it has to use `from_numpy` instead.

On the other hand, I don't think that `completely_flatten` can return a NumPy MaskedArray. Nope, it can't ([v1 implementation](https://github.com/scikit-hep/awkward-1.0/blob/4071495c93fe50443827bbc286ace45ae5fa7ba0/src/awkward/_util.py#L578-L631), which is all in one place).

It should work if you just click this suggestion in. The code will be more maintainable if it's not checking for impossible cases.

```suggestion
        out = ak._v2.contents.NumpyArray(nplike.concatenate(out))
```